File: src-tauri/src/commands/mqtt.rs
use crate::error::AppError;
use crate::state::AppState;
use crate::mqtt::types::*;
use tauri::State;

// ─── Connection Commands ─────────────────────────────────────────

#[tauri::command]
pub async fn mqtt_connect(
    state: State<'_, AppState>,
    config: MqttConnectionConfig,
) -> Result<String, AppError> {
    state.mqtt_manager.connect(config).await
}

#[tauri::command]
pub async fn mqtt_disconnect(
    state: State<'_, AppState>,
    connection_id: String,
) -> Result<(), AppError> {
    state.mqtt_manager.disconnect(&connection_id).await
}

#[tauri::command]
pub async fn mqtt_get_connections(
    state: State<'_, AppState>,
) -> Result<Vec<MqttConnectionInfo>, AppError> {
    Ok(state.mqtt_manager.list_connections().await)
}

#[tauri::command]
pub async fn mqtt_get_connection_status(
    state: State<'_, AppState>,
    connection_id: String,
) -> Result<MqttConnectionStatus, AppError> {
    state.mqtt_manager.get_status(&connection_id).await
}

// ─── Subscription Commands ───────────────────────────────────────

#[tauri::command]
pub async fn mqtt_subscribe(
    state: State<'_, AppState>,
    connection_id: String,
    request: MqttSubscribeRequest,
) -> Result<MqttSubscriptionInfo, AppError> {
    state
        .mqtt_manager
        .subscribe(&connection_id, &request)
        .await
}

#[tauri::command]
pub async fn mqtt_unsubscribe(
    state: State<'_, AppState>,
    connection_id: String,
    subscription_id: u32,
) -> Result<(), AppError> {
    state
        .mqtt_manager
        .unsubscribe(&connection_id, subscription_id)
        .await
}

#[tauri::command]
pub async fn mqtt_get_subscriptions(
    state: State<'_, AppState>,
    connection_id: String,
) -> Result<Vec<MqttSubscriptionInfo>, AppError> {
    state
        .mqtt_manager
        .get_subscriptions(&connection_id)
        .await
}

// ─── Publish Commands ────────────────────────────────────────────

#[tauri::command]
pub async fn mqtt_publish(
    state: State<'_, AppState>,
    connection_id: String,
    request: MqttPublishRequest,
) -> Result<(), AppError> {
    state
        .mqtt_manager
        .publish(&connection_id, &request)
        .await
}

// ─── Poll Commands ───────────────────────────────────────────────

#[tauri::command]
pub async fn mqtt_poll_messages(
    state: State<'_, AppState>,
    connection_id: String,
) -> Result<MqttPollResponse, AppError> {
    state.mqtt_manager.poll_messages(&connection_id).await
}

// ─── Topic Commands ──────────────────────────────────────────────

#[tauri::command]
pub async fn mqtt_get_topics(
    state: State<'_, AppState>,
    connection_id: String,
) -> Result<Vec<MqttTopicInfo>, AppError> {
    state.mqtt_manager.get_topics(&connection_id).await
}

// ─── Broker Admin Commands ───────────────────────────────────────

#[tauri::command]
pub async fn mqtt_get_broker_stats(
    state: State<'_, AppState>,
    connection_id: String,
) -> Result<BrokerStats, AppError> {
    state
        .mqtt_manager
        .get_broker_stats(&connection_id)
        .await
}

#[tauri::command]
pub async fn mqtt_get_broker_clients(
    state: State<'_, AppState>,
    connection_id: String,
) -> Result<Vec<BrokerClientInfo>, AppError> {
    state
        .mqtt_manager
        .get_broker_clients(&connection_id)
        .await
}

File: src/services/mqtt.ts
import { invoke } from "@tauri-apps/api/core";
import { withLogging } from "@/services/logger";
import type {
  MqttConnectionConfig,
  MqttConnectionInfo,
  MqttConnectionStatus,
  MqttSubscribeRequest,
  MqttSubscriptionInfo,
  MqttPublishRequest,
  MqttPollResponse,
  MqttTopicInfo,
  BrokerStats,
  BrokerClientInfo,
} from "@/types/mqtt";

// ─── Connection ──────────────────────────────────────────────────

export const mqttConnect = withLogging(
  "mqtt_connect",
  async (config: MqttConnectionConfig): Promise<string> => {
    return invoke("mqtt_connect", { config });
  }
);

export const mqttDisconnect = withLogging(
  "mqtt_disconnect",
  async (connectionId: string): Promise<void> => {
    return invoke("mqtt_disconnect", { connectionId });
  }
);

export const mqttGetConnections = withLogging(
  "mqtt_get_connections",
  async (): Promise<MqttConnectionInfo[]> => {
    return invoke("mqtt_get_connections");
  }
);

export const mqttGetConnectionStatus = withLogging(
  "mqtt_get_connection_status",
  async (connectionId: string): Promise<MqttConnectionStatus> => {
    return invoke("mqtt_get_connection_status", { connectionId });
  }
);

// ─── Subscriptions ───────────────────────────────────────────────

export const mqttSubscribe = withLogging(
  "mqtt_subscribe",
  async (connectionId: string, request: MqttSubscribeRequest): Promise<MqttSubscriptionInfo> => {
    return invoke("mqtt_subscribe", { connectionId, request });
  }
);

export const mqttUnsubscribe = withLogging(
  "mqtt_unsubscribe",
  async (connectionId: string, subscriptionId: number): Promise<void> => {
    return invoke("mqtt_unsubscribe", { connectionId, subscriptionId });
  }
);

export const mqttGetSubscriptions = withLogging(
  "mqtt_get_subscriptions",
  async (connectionId: string): Promise<MqttSubscriptionInfo[]> => {
    return invoke("mqtt_get_subscriptions", { connectionId });
  }
);

// ─── Publish ─────────────────────────────────────────────────────

export const mqttPublish = withLogging(
  "mqtt_publish",
  async (connectionId: string, request: MqttPublishRequest): Promise<void> => {
    return invoke("mqtt_publish", { connectionId, request });
  }
);

// ─── Messages & Topics ──────────────────────────────────────────

export const mqttPollMessages = withLogging(
  "mqtt_poll_messages",
  async (connectionId: string): Promise<MqttPollResponse> => {
    return invoke("mqtt_poll_messages", { connectionId });
  }
);

export const mqttGetTopics = withLogging(
  "mqtt_get_topics",
  async (connectionId: string): Promise<MqttTopicInfo[]> => {
    return invoke("mqtt_get_topics", { connectionId });
  }
);

// ─── Broker Admin ────────────────────────────────────────────────

export const mqttGetBrokerStats = withLogging(
  "mqtt_get_broker_stats",
  async (connectionId: string): Promise<BrokerStats> => {
    return invoke("mqtt_get_broker_stats", { connectionId });
  }
);

export const mqttGetBrokerClients = withLogging(
  "mqtt_get_broker_clients",
  async (connectionId: string): Promise<BrokerClientInfo[]> => {
    return invoke("mqtt_get_broker_clients", { connectionId });
  }
);
